lib/jobs/switch-node-relations.js

```diff
@@ -0,0 +1,212 @@
+    waterline,
+    catalogSearch,
+    Logger,
+    util,
+    Promise,
+    assert,
+    _
+) {
+
+    var logger = Logger.initialize(switchRelationsJobFactory);
+
+    /**
+     *
+     * @param {Object} [options]
+     * @constructor
+     */
+    function SwitchRelationsJob(options, context, taskId) {
+        SwitchRelationsJob.super_.call(this, logger, options, context, taskId);
+
+        this.nodeId = context.target || options.nodeId;
+        assert.isMongoId(this.nodeId);
+    }
+
+    util.inherits(SwitchRelationsJob, BaseJob);
+
+    /**
+     * Globals
+     */
+
+    var RELATION_TYPE = 'connectsTo';
+
+    /**
+     * @memberOf SwitchRelationsJob
+     */
+    SwitchRelationsJob.prototype._run = function run() {
+        var self = this;
+        var lldpList;
+        return self._getMacAddressList()
+        .then(function (macList) {
+            lldpList = macList;
+            return _getCatalogsBySource('lldp');
+        })
+        .each(function (lldpData) {
+            return _findMatchingMac(lldpData, lldpList)
+            .then(function (info) {
+                if (Object.keys(info).length) {
+                    return self._updateSwitchRelations(lldpData, info);
+                }
+            });
+        })
+        .then(function () {
+            self._done();
+        })
+        .catch(function (err) {
+            self._done(err);
+        });
+
+    };
+
+    /**
+     * search in snmp-1 catalog data for switch mac addresses
+     * returns an array of switch mac addresses
+     */
+    SwitchRelationsJob.prototype._getMacAddressList = function getMacAddressList() {
+        var self = this;
+        var data;
+        var macArr = [];
+        return waterline.catalogs.findMostRecent({
+            node: self.nodeId,
+            source: 'snmp-1'
+        })
+        .then(function (catalog) {
+            data = catalog.data;
+            return Promise.resolve(Object.keys(catalog.data));
+        })
+        .filter(function (key) {
+            return /ifPhysAddress_/.test(key);
+        })
+        .each(function (phys) {
+            if (_validateMacStrFormat(data[phys])) {
+                macArr.push(_adjustMacStrFormat(data[phys]));
+            }
+        })
+        .then(function () {
+            if (macArr.length === 0) {
+                throw new Error('couldnt find ports mac addresses in catalog');
+            }
+            return Promise.resolve(macArr);
+        });
+    };
+
+    /**
+     * update an existing entry in nodes relations or
+     * create one if it doesn't exist
+     * return a promise
+     */
+    SwitchRelationsJob.prototype._updateSwitchRelations =
+        function updateSwitchRelations(lldpData, info) {
+            var self = this;
+            return waterline.nodes.findByIdentifier(lldpData.node)
+            .then(function (node) {
+                if (!node) {
+                    return Promise.reject('Could not find node with identifier ' + lldpData.node);
```

Comment:
@keedya This will work fine as is, but just FYI, you can also throw errors inside a `.then` callback.  These will be converted to rejections automatically.  The same is true when you want a Promise to fulfill with a specific value.  You can just return values directly without wrapping in `Promise.resolve`.  Generally you only need to return a Promise from a `.then` callback if that promise is pending an you need to wait for it to fulfill or reject prior to executing the next callback in the chain.